packages/dht/src/dht/DhtNode.ts

```diff
@@ -0,0 +1,629 @@
+import { ConnectionManager } from '../connection/ConnectionManager'
+import { DhtRpcClient } from '../proto/DhtRpc.client'
+import { Logger } from '../helpers/Logger'
+import { v4 } from 'uuid'
+import { jsFormatPeerDescriptor } from '../helpers/common'
+import { IDhtRpc } from '../proto/DhtRpc.server'
+
+export interface RouteMessageParams {
+    message: Uint8Array
+    destinationPeer: PeerDescriptor
+    sourcePeer: PeerDescriptor
+    appId: string
+    previousPeer?: PeerDescriptor
+    messageId?: string
+}
+
+export enum Event {
+    NEW_CONTACT = 'streamr:dht:dht-node:new-peer',
+    CONTACT_REMOVED = 'streamr:dht:dht-node:peer-removed',
+    JOIN_COMPLETED = 'streamr:dht:dht-node:join-completed'
+}
+
+export const DEFAULT_APP_ID = 'layer0'
+
+export interface DhtNodeConfig {
+    transportLayer?: ITransport
+    peerDescriptor?: PeerDescriptor
+    entryPoints?: PeerDescriptor[]
+    webSocketHost?: string
+    webSocketPort?: number
+    peerIdString?: string
+    appId?: string
+    numberOfNodesPerKBucket?: number,
+    nodeName?: string
+}
+
+const logger = new Logger(module)
+
+export class DhtNode extends EventEmitter implements ITransport, IDhtRpc {
+    private noProgressCounter = 0
+    private readonly PARALLELISM = 3
+    private readonly MAX_NEIGHBOR_LIST_SIZE = 100
+    private readonly NUMBER_OF_NODES_PER_K_BUCKET = 1
+    private readonly JOIN_NO_PROGRESS_LIMIT = 4
+    private readonly peers: Map<string, DhtPeer>
+    private readonly numberOfNodesPerKBucket: number
+    private readonly routerDuplicateDetector: RouterDuplicateDetector
+    private readonly appId: string
+    private readonly ongoingClosestPeersRequests: Set<string>
+    private joinTimeoutRef: NodeJS.Timeout | null = null
+    private ongoingJoinOperation = false
+
+    private bucket?: KBucket<DhtPeer>
+    private neighborList?: SortedContactList
+    private openInternetPeers?: SortedContactList
+    private rpcCommunicator?: RoutingRpcCommunicator
+    private transportLayer?: ITransport
+    private ownPeerDescriptor?: PeerDescriptor
+    private ownPeerId?: PeerID
+
+    private outgoingClosestPeersRequestsCounter = 0
+
+    private cleanUpHandleForConnectionManager?: ConnectionManager
+    private started = false
+    private stopped = false
+
+    constructor(private config: DhtNodeConfig) {
+        super()
+
+        this.peers = new Map()
+
+        this.appId = config.appId ?? DEFAULT_APP_ID
+
+        this.numberOfNodesPerKBucket = config.numberOfNodesPerKBucket || this.NUMBER_OF_NODES_PER_K_BUCKET
+        this.ongoingClosestPeersRequests = new Set()
+        this.routerDuplicateDetector = new RouterDuplicateDetector()
+    }
+
+    public async start(): Promise<void> {
+        if (this.started || this.stopped) {
+            return
+        }
+        logger.info(`Starting new Streamr Network DHT Node on ${this.appId === DEFAULT_APP_ID ? 'Layer 0' : 'Layer 1 (stream)'}`)
```

Comment:
The `Layer 1 (stream)` value can be misleading if a user defines a custom appId. 

Most likely I don't understand the layering well enough. But would it make sense to give these layers some human-readable names, e.g.:
- `layer0` = network layer
- `layer1` = stream layer

That kind of name would be easy to understand and would already have some semantics attached to it.

Also, is the control layer a combination of all layers? Or is that actually something else than layer?

Reply:
The whole conditional expression has been removed